### PR TITLE
 [Php80] Add MixedTypeRector 

### LINKED ISSUE #7301: [PHP 8.0] Add inline mixed type rule
Since PHP 8.0, the mixed type can be easily inlined. Apart the parent contract method from interface/abstract class, this cannot be BC break, as type declaration accepts any type:

* RFC: https://wiki.php.net/rfc/mixed_type_v2

```diff
-/**
- * @param mixed $element
- */
-private function isRectorClass($element): bool
+private function isRectorClass(mixed $element): bool
 {
 }
```

### PR DESCRIPTION
**Pull request by samsonasik**
Fixes https://github.com/rectorphp/rector/issues/7301

### CODE REVIEW

File: rules/Arguments/ArgumentDefaultValueReplacer.php [hunk 1]
```diff
@@ -68,9 +66,9 @@ public function isDefaultValueMatched(?Expr $expr, $value): bool
     }
 
     private function processParams(
-        ClassMethod $classMethod,
-        ReplaceArgumentDefaultValueInterface $replaceArgumentDefaultValue
-    ): ?ClassMethod {
+        object $classMethod,
+        \ReplaceArgumentDefaultValueInterface $replaceArgumentDefaultValue
```
Comment (samsonasik) [inline rules/Arguments/ArgumentDefaultValueReplacer.php]: it seems may cause invalid replacement:

```diff
 1/1 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
1 file with changes
===================

1) rules/Arguments/ArgumentDefaultValueReplacer.php:3

    ---------- begin diff ----------
@@ @@

 namespace Rector\Arguments;

+use object;
 use PhpParser\BuilderHelpers;
 use PhpParser\Node;
 use PhpParser\Node\Arg;
@@ @@
         return $this->processArgs($node, $replaceArgumentDefaultValue);
     }

-    /**
-     * @param mixed $value
-     */
-    public function isDefaultValueMatched(?Expr $expr, $value): bool
+    public function isDefaultValueMatched(?Expr $expr, mixed $value): bool
     {
         // allow any values before, also allow param without default value
         if ($value === ReplaceArgumentDefaultValue::ANY_VALUE_BEFORE) {
@@ @@
     }

     private function processParams(
-        ClassMethod $classMethod,
-        ReplaceArgumentDefaultValueInterface $replaceArgumentDefaultValue
-    ): ?ClassMethod {
+        object $classMethod,
+        \ReplaceArgumentDefaultValueInterface $replaceArgumentDefaultValue
+    ): ?object {
         $position = $replaceArgumentDefaultValue->getPosition();

         if (! $this->isDefaultValueMatched(
@@ @@
    ----------- end diff -----------

Applied rules:
 * CountOnNullRector (https://3v4l.org/Bndc9)
 * MixedTypeRector
 * AddMethodCallBasedStrictParamTypeRector (https://github.com/symplify/phpstan-rules/blob/master/docs/rules_overview.md#checktypehintcallertyperule)
 * ReturnTypeDeclarationRector (https://wiki.php.net/rfc/scalar_type_hints_v5)
```

Comment (samsonasik) [reply rules/Arguments/ArgumentDefaultValueReplacer.php]: @TomasVotruba it seems there is a bug in `IntersectionTypeMapper` which somehow replace to `object`

https://github.com/rectorphp/rector-src/blob/764416672a588c4bd80084a853973773f9835690/packages/PHPStanStaticTypeMapper/TypeMapper/IntersectionTypeMapper.php#L96-L98

Comment (samsonasik) [reply rules/Arguments/ArgumentDefaultValueReplacer.php]: The issue seems when combined with `AddMethodCallBasedStrictParamTypeRector`

File: rules/Arguments/ArgumentDefaultValueReplacer.php [hunk 2]
```diff
@@ -166,7 +159,7 @@ private function processArrayReplacement(
      */
     private function resolveArgumentValuesToBeforeRecipe(
         array $argumentNodes,
-        ReplaceArgumentDefaultValueInterface $replaceArgumentDefaultValue
+        \ReplaceArgumentDefaultValueInterface $replaceArgumentDefaultValue
```
Comment (samsonasik) [inline rules/Arguments/ArgumentDefaultValueReplacer.php]: debuggin: 

```php
$type = $this->staticTypeMapper->mapPhpParserNodePHPStanType($phpParserTypeNode);
 dump($type);
 ```

 on `ClassMethodParamTypeCompleter` seems got type:

```php
Rector\StaticTypeMapper\ValueObject\Type\FullyQualifiedObjectType #78514
   subtractedType: null
   cachedParent: null
   cachedInterfaces: null
   currentAncestors: array (0)
   className: 'ReplaceArgumentDefaultValueInterface'
   classReflection: null
```

which namespaced in use not included.

https://github.com/rectorphp/rector-src/blob/1836de97836c78ae4c85c1fdce3664688fb1011f/rules/TypeDeclaration/NodeAnalyzer/ClassMethodParamTypeCompleter.php#L63

Comment (samsonasik) [reply rules/Arguments/ArgumentDefaultValueReplacer.php]: tweak by check virtual node https://github.com/rectorphp/rector-src/pull/2701/commits/39e3969a2a4884cf6bceeee558f36ceef4f4c5d0

File: rules/Arguments/ArgumentDefaultValueReplacer.php (hunk 1 above)
Comment (samsonasik) [reply rules/Arguments/ArgumentDefaultValueReplacer.php]: tweak by check virtual node https://github.com/rectorphp/rector-src/pull/2701/commits/39e3969a2a4884cf6bceeee558f36ceef4f4c5d0

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

**For summary:**

- Applied mixed type when:
   - no param type
   - param doc is mixed

- Handle virtual Node when combined with other rules https://github.com/rectorphp/rector-src/pull/2701/files#diff-42a5a059b66868f3ede960ea9a50a3591de610668457d76063ac2f6dc653dcc3 , ref 
   - https://github.com/rectorphp/rector-src/pull/2701#discussion_r927432461
   - https://github.com/rectorphp/rector-src/pull/2701#discussion_r927707115

For Return Type, I think that require different PR as that will need various handling, eg: actually return `void`.

File: src/PhpParser/Printer/BetterStandardPrinter.php [hunk 3]
```diff
@@ -149,7 +149,7 @@ public function pFileWithoutNamespace(FileWithoutNamespace $fileWithoutNamespace
         return ltrim($content);
     }
 
-    protected function p(Node $node, $parentFormatPreserved = false): string
+    protected function p(Node $node, mixed $parentFormatPreserved = false): string
```
Comment (samsonasik) [inline src/PhpParser/Printer/BetterStandardPrinter.php]: @TomasVotruba this possibly need to be skipped? the parent `p` method has signature:

```php
     * @param Node $node Node to be pretty printed
     * @param bool $parentFormatPreserved Whether parent node has preserved formatting
     *
     * @return string Pretty printed node
     */
    protected function p(Node $node, $parentFormatPreserved = false) : string {
```

ref https://github.com/nikic/PHP-Parser/blob/de4ac93023427c778eecc5ce14d194f1b33c8cd6/lib/PhpParser/PrettyPrinterAbstract.php#L513-L518

Comment (samsonasik) [reply src/PhpParser/Printer/BetterStandardPrinter.php]: I think to reduce unwanted change, the solution is to skip if no `@param mixed` doc https://github.com/rectorphp/rector-src/pull/2701/commits/a569a4b410fbbc9d24527e169f7f76747a16745e

**Comment by samsonasik** [comment]
it seems when parent method has `mixed` type already, and current doesn't has type doc, it should be transformed to `mixed`.

**Comment by samsonasik** [comment]
@TomasVotruba so, I think the use case like this should be skipped:

```php
class VendorParam
{
    /**
     * @param int $param
     */
    public function run($param)
    {}
}

class UseParam extends VendorParam
{
    public function run($param)
    {}
}
```

Next, when parent method `int` param doc be typed, has child widen with mixed should not be applied while it is working ok, same type is better.

There are other use case like typed already that should be skipped as well:

```php
class A
{
    public function run(int $a)
    {}
}

class B extends A
{
    public function run($a)
    {}
}
```

as same type in child should be preferred, so there should be different rule or phpstan that handle that 

https://phpstan.org/r/c1ffffc4-e76a-4367-adf7-26608665d9b1

**Comment by samsonasik** [comment]
implemented with add `MakeClassMethodParamMixedTypedGuard`

**Comment by TomasVotruba** [comment]
> @TomasVotruba so, I think the use case like this should be skipped:

Agreed. I think only the doc-typed `mixed` should be included, so the intention is already defined by user.
Implicit mixed types should be skipped, as the absence of doc-type does not mean mixed type. Maybe it's just missing.

```diff
public function run($value)
{
}

-/**
- * @param mixed $value
- */
-public function run($value)
+public function run(mixed $value)
 {
 }
```

File: rules/CodingStyle/ClassNameImport/ShortNameResolver.php [hunk 4]
```diff
@@ -171,7 +171,7 @@ private function resolveFromStmtsDocBlocks(array $stmts): array
             $phpDocNodeTraverser->traverseWithCallable(
                 $phpDocInfo->getPhpDocNode(),
                 '',
-                static function ($node) use (&$shortNames) {
+                static function (mixed $node) use (&$shortNames) {
```
Comment (samsonasik) [inline rules/CodingStyle/ClassNameImport/ShortNameResolver.php]: @TomasVotruba it seems it should be skipped as used as `Arg`, and the signature originally as callable `callable(Node $node)`:

```php
    /**
     * @param callable(Node $node): (int|null|Node) $callable
     */
    public function traverseWithCallable(Node $node, string $docContent, callable $callable): void
```

**Comment by samsonasik** [comment]
The issue with only check `mixed` doc type is when we use the following code in code base:

```php
interface SomeInterface 
{ 
    /** @param mixed $data */
    public function save($data);
}

class Memory extends SomeInterface
{
    public function save($data) {}
}
```

That will only change interface, and that will cause error something like:

```bash

 ------------------------------------------------------------------------
  packages/Caching/ValueObject/Storage/MemoryCacheStorage.php:37
 ------------------------------------------------------------------------
  - '#Method Rector\\Caching\\ValueObject\\Storage\\MemoryCacheStorage\:\:save\(\) has parameter \$data with no type specified#'
```

see https://github.com/rectorphp/rector-src/runs/7473011787?check_suite_focus=true#step:5:16 , that's why I created `MakeClassMethodParamMixedTypedGuard` to cover that.

**Comment by samsonasik** [comment]
Ok, let's [only make typed with use doc](https://github.com/rectorphp/rector-src/pull/2701/commits/8b638ccbe8d93c0f4031fc1f95936068aa1b2c7b) @param [mixed](https://github.com/rectorphp/rector-src/pull/2701/commits/8b638ccbe8d93c0f4031fc1f95936068aa1b2c7b)

**Comment by TomasVotruba** [comment]
I think these cases should be skipped, as imlicit mixed doc can be overriden by explicit type in of childs:

Then it could crash:
https://3v4l.org/DGidN

**Comment by samsonasik** [comment]
@TomasVotruba It seems we need to ensure it only apply change when:

- no parent method
- no child method

that will also ensure no manual add `mixed` type when parent has `@param mixed` and child not, or child using `@inheritDoc`, what do you think?

**Comment by TomasVotruba** [comment]
That makes sense :+1:

**Comment by samsonasik** [comment]
for ClassMethod, I think the mixed type can be applied for final class.

**Comment by samsonasik** [comment]
or not, as final class can have parent

**Comment by samsonasik** [comment]
implemented check has child or parent method :tada:

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

**Comment by samsonasik** [comment]
rebased.

**Comment by TomasVotruba** [comment]
Thanks :+1:

**Pull request merged**